### PR TITLE
Tests for reconnection logic

### PR DESCRIPTION
**Pull request by jhorstmann**

### CODE REVIEW

File: src/main/java/org/zalando/fahrschein/NakadiReader.java [hunk 1]
```diff
@@ -192,11 +193,11 @@ private void readEvent(final Iterator<T> source, final List<T> target) throws Js
         }
     }
 
-    public void run() throws IOException {
+    public void run() throws IOException, BackoffException {
```
Comment (lukasniemeier-zalando) [inline src/main/java/org/zalando/fahrschein/NakadiReader.java]: What about catching the backoff exceptions in here and then only expose the `IOException`?

File: src/main/java/org/zalando/fahrschein/NakadiReader.java [hunk 2]
```diff
@@ -268,15 +266,17 @@ public void run(long timeout, TimeUnit timeoutUnit) throws IOException {
     private void expectField(JsonParser jsonParser, String expectedFieldName) throws IOException {
         final String fieldName = jsonParser.getCurrentName();
         if (fieldName == null) {
-            throw new IOException("Stream was closed or no field at current position");
+            throw new EOFException("Stream was closed or no field at current position");
```
Comment (lukasniemeier-zalando) [inline src/main/java/org/zalando/fahrschein/NakadiReader.java]: Can this actually happen here?

File: src/main/java/org/zalando/fahrschein/NoBackoffStrategy.java [hunk 3]
```diff
@@ -6,14 +6,6 @@
 public class NoBackoffStrategy implements BackoffStrategy{
     @Override
     public <T> T call(final int initialExceptionCount, final @Nullable IOException initialException, final IOCallable<T> callable) throws BackoffException, InterruptedException {
```
Comment (lukasniemeier-zalando) [inline src/main/java/org/zalando/fahrschein/NoBackoffStrategy.java]: initialException not nullable?

File: src/test/java/org/zalando/fahrschein/NakadiReaderTest.java [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+
+        final NoBackoffStrategy backoffStrategy = new NoBackoffStrategy();
+
+        final NakadiReader<SomeEvent> nakadiReader = new NakadiReader<>(uri, clientHttpRequestFactory, backoffStrategy, cursorManager, objectMapper, EVENT_NAME, Optional.<Subscription>empty(), SomeEvent.class, listener);
+
+        try {
+            nakadiReader.run();
+            fail("Expected IOException on initial connect");
+        } catch (IOException e) {
+            assertEquals("Initial connection failed", e.getMessage());
```
Comment (lukasniemeier-zalando) [inline src/test/java/org/zalando/fahrschein/NakadiReaderTest.java]: Change to `expectedException` `Rule` with `expectMessage`?

File: src/test/java/org/zalando/fahrschein/NakadiReaderTest.java [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+        }
+    }
+
+    @Test
+    public void shouldNotReconnectWithoutBackoff() throws IOException, InterruptedException, BackoffException {
+        final ClientHttpResponse response = mock(ClientHttpResponse.class);
+        //final ByteArrayInputStream initialInputStream = new ByteArrayInputStream("{\"cursor\":{\"partition\":\"0\",\"offset\":\"0\"}}".getBytes("utf-8"));
```
Comment (lukasniemeier-zalando) [inline src/test/java/org/zalando/fahrschein/NakadiReaderTest.java]: remove?

**Comment by lukasniemeier-zalando** [comment]
You also could integrate the ideas from #32 :wink:

File: src/main/java/org/zalando/fahrschein/NakadiReader.java [hunk 6]
```diff
@@ -197,10 +199,20 @@ public void run() throws IOException {
     }
 
     public void run(long timeout, TimeUnit timeoutUnit) throws IOException {
+        try {
+            runInternal(timeout, timeoutUnit);
+        } catch (BackoffException e) {
+            LOG.warn("");
```
Comment (lukasniemeier-zalando) [inline src/main/java/org/zalando/fahrschein/NakadiReader.java]: uh ha :wine_glass:

File: src/main/java/org/zalando/fahrschein/NakadiReader.java [hunk 7]
```diff
@@ -236,25 +248,26 @@ public void run(long timeout, TimeUnit timeoutUnit) throws IOException {
...
+                    LOG.debug("Reconnecting after [{}] errors", errorCount);
                     jsonInput = backoffStrategy.call(errorCount, e, this::openJsonInput);
                     jsonParser = jsonInput.getJsonParser();
                     LOG.info("Reconnected after [{}] errors", errorCount);
-                } catch (BackoffException e1) {
-                    LOG.warn("Could not reconnect after [{}] errors", errorCount, e1);
-                    return;
                 } catch (InterruptedException e1) {
-                    LOG.warn("Interrupted during reconnection");
+                    LOG.warn("Interrupted during reconnection", e);
```
Comment (lukasniemeier-zalando) [inline src/main/java/org/zalando/fahrschein/NakadiReader.java]: `e` vs `e1`

Comment (jhorstmann) [inline src/main/java/org/zalando/fahrschein/NakadiReader.java]: > e vs e1

Good catch

**Comment by lukasniemeier-zalando** [comment]
:+1:

**Pull request merged**